Sources/SwiftProtobuf/Google_Protobuf_Any+Registry.swift

```diff
@@ -0,0 +1,100 @@
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+// This should be deleted once https://github.com/apple/swift-protobuf/pull/377
+// lands as directly using `protoMessageName` will be all that is needed.
+internal func messageName(fromMessageType messageType: Message.Type) -> String {
+  let protoPackageName = messageType.protoPackageName
+  if protoPackageName.isEmpty {
+    return messageType.protoMessageName
+  } else {
+    return "\(protoPackageName).\(messageType.protoMessageName)"
+  }
+}
+
+fileprivate let wktToMessageType: [String:Message.Type] = [
+  "google.protobuf.Any": Google_Protobuf_Any.self,
+  "google.protobuf.BoolValue": Google_Protobuf_BoolValue.self,
+  "google.protobuf.BytesValue": Google_Protobuf_BytesValue.self,
+  "google.protobuf.DoubleValue": Google_Protobuf_DoubleValue.self,
+  "google.protobuf.Duration": Google_Protobuf_Duration.self,
+  "google.protobuf.FieldMask": Google_Protobuf_FieldMask.self,
+  "google.protobuf.FloatValue": Google_Protobuf_FloatValue.self,
+  "google.protobuf.Int32Value": Google_Protobuf_Int32Value.self,
+  "google.protobuf.Int64Value": Google_Protobuf_Int64Value.self,
+  "google.protobuf.ListValue": Google_Protobuf_ListValue.self,
+  "google.protobuf.StringValue": Google_Protobuf_StringValue.self,
+  "google.protobuf.Struct": Google_Protobuf_Struct.self,
+  "google.protobuf.Timestamp": Google_Protobuf_Timestamp.self,
+  "google.protobuf.UInt32Value": Google_Protobuf_UInt32Value.self,
+  "google.protobuf.UInt64Value": Google_Protobuf_UInt64Value.self,
+  "google.protobuf.Value": Google_Protobuf_Value.self,
+]
+
+fileprivate var knownTypes = [String:Message.Type]()
+
+public extension Google_Protobuf_Any {
+
+    /// Register a message type so that Any objects can use
+    /// them for decoding contents.
+    ///
+    /// This is currently only required in two cases:
+    ///
+    /// * When decoding Protobuf Text format.  Currently,
+    ///   Any objects do not defer deserialization from Text
+    ///   format.  Depending on how the Any objects are stored
+    ///   in text format, the Any object may need to look up
+    ///   the message type in order to deserialize itself.
+    ///
+    /// * When re-encoding an Any object into a different
+    ///   format than it was decoded from.  For example, if
+    ///   you decode a message containing an Any object from
+    ///   JSON format and then re-encode the message into Protobuf
+    ///   Binary format, the Any object will need to complete the
+    ///   deferred deserialization of the JSON object before it
+    ///   can re-encode.
+    ///
+    /// Note that well-known types are pre-registered for you and
+    /// you do not need to register them from your code.
+    ///
+    /// Also note that this is not needed if you only decode and encode
+    /// to and from the same format.
+    @discardableResult static public func register(messageType: Message.Type) {
```

Comment:
Oops, was playing with making it return a bool to cover collisions when I wondered about the problem now in 379.